### ISSUE TITLE
Migrate Command & Directories

### ISSUE THREAD

**Issue by jMartich** [bsab/django-unifi-portal#4]
Good Morning Everyone, 
Excuse my lack of knowledge here, but I can't seem to put together the step between installing the unifi-portal & the migrate command. I have already installed the unifi-portal using pip; just wondering if the "migrate" command supposed to just work right after? Or do I have to setup the environment variables for django before I can issue that command? Also reading further, there's mention of settings.py file, what directory does that go in? I'm just asking in advance. Finally, is this something I can run on an apache server with virtual hosts or should I just use a dedicated machine? Thank you for your time guys.

**Comment by KevSex** [comment]
I had issues with the migrate command too. I never ended up using it and used an alternative method.
Have you set up your virtual environment and the django project? 

This is my first django app, I've got it set up but been meaning to delete it and start from scratch to get my head around it more. Just need to find the time. Will try document as I go.

**Comment by jMartich** [comment]
Hello KevSex, funny you mention virtual environment; I just deployed an Unbuntu 16 VM to install DJango from scratch to get more familiar with it as well. I'm willing to document more as well for the sake of this project. At the moment I'm running the custom wifi portal that's mentioned on this project on an AWS instance for a customer but it's not working as I thought it would. This project is exactly what I need for my customer, but I still need a tad bit more experience with DJango. I don't know what timezone you're in but if you'd like to collaborate a little with we me on my endeavor perhaps we can learn from each other. Thanks again for your response. Also sorry to be such a noob, but is there a way to PM on here?

To Clarify, the VM is for me to run virtualenv.

**Comment by jMartich** [comment]
As I mentioned before I installed django on a fresh VM using this [guide](https://www.digitalocean.com/community/tutorials/how-to-install-django-and-set-up-a-development-environment-on-ubuntu-16-04) to get more familiar with the way django works. I got virtualenv working properly, tested the test project, but still can't seem to get the migrate command to work the way it is documented on here. For the record I'm using python3 along with pip3 which I am sure have something or everything to do with my attempts not working properly. Here's the output I get now from the VM:

jmartich@Ubuntu-VM01:~/Projects/Unifi$ migrate django_unifi_portal
Usage: migrate COMMAND ...

    Available commands:
        compare_model_to_db          - compare MetaData against the current database state
	create                       - create an empty repository at the specified path
	create_model                 - dump the current database as a Python model to stdout
	db_version                   - show the current version of the repository under version control
	downgrade                    - downgrade a database to an earlier version
	drop_version_control         - removes version control from a database
	help                         - displays help on a given command
	make_update_script_for_model - create a script changing the old MetaData to the new (current) MetaData
	manage                       - creates a Python script that runs Migrate with a set of default values
	script                       - create an empty change Python script
	script_sql                   - create empty change SQL scripts for given database
	source                       - display the Python code for a particular version in this repository
	test                         - performs the upgrade and downgrade command on the given database
	update_db_from_model         - modify the database to match the structure of the current MetaData
	upgrade                      - upgrade a database to a later version
	version                      - display the latest version available in a repository
	version_control              - mark a database as under this repository's version control

    Enter "migrate help COMMAND" for information on a particular command.
    

migrate: error: Invalid command django_unifi_portal

Here's the output I get from pip3 freeze:

jmartich@Ubuntu-VM01:~/Projects/Unifi$ pip3 freeze
chardet==2.3.0
command-not-found==0.3
decorator==4.0.6
defusedxml==0.5.0
Django==2.0
django-braces==1.12.0
django-material==1.1.1
django-oauth-toolkit==1.0.0
django-rest-framework-social-oauth2==1.0.8
django-unifi-portal==0.0.2
djangorestframework==3.7.3
language-selector==0.1
oauthlib==2.0.6
olefile==0.44
pbr==1.8.0
Pillow==4.3.0
pycurl==7.43.0
pygobject==3.20.0
PyJWT==1.5.3
python-apt==1.1.0b1
python-debian==0.1.27
python-systemd==231
python3-openid==3.1.0
pytz==2017.3
requests==2.9.1
requests-oauthlib==0.8.0
requests-toolbelt==0.8.0
six==1.10.0
social-auth-app-django==2.0.0
social-auth-core==1.5.0
SQLAlchemy==1.0.11
sqlalchemy-migrate==0.10.0
sqlparse==0.1.18
ssh-import-id==5.5
Tempita==0.5.2
ufw==0.35
unattended-upgrades==0.1
urllib3==1.13.1
virtualenv==15.1.0

Please let me know if I am missing anything. Thanks in advance everyone.

**Comment by KevSex** [comment]
@jMartich 
Hey, sorry been a busy day! I'm in the UK. I'm also a noob with github but doesn't look like there is a PM feature. :/ Happy to assist as much as I can to help you out though.

First off, I too ran into the issue that you look to have. I don't think this project supports python 3. I got too many errors when attempting to run so switched to python 2.7 and looks to be all good.

I've got the same setup, AWS EC2 instance running Ubuntu 16.04. Looking back at my history, within the virtualenv, I've used pip to install django then the same for django-unifi-portal. (django will attempt to install the latest version which isn't supported in python 2.7 therefore you'll need to run `pip install django==1.11.7`).Within the project, I've then downloaded this git repo then I think I ran 'python manage.py migrate'. 

I'm going to spin up another EC2 instance and start fresh. Will let you know how it goes.

**Comment by bsab** [comment]
@jMartich 
to make the migrations do:
python manage.py migrate

**Comment by KevSex** [comment]
Some more info that I found when setting this up, for step 3 in the Quick start, I had to remove this as it complained about a duplicate being found.
Also for step 6, adding in the urls line, I had to add in `include` to the line `from django.conf.urls import`

**Comment by jMartich** [comment]
Thanks @bsab & @KevSex, I'm going to spin up a new VM today & use python 2.7. The reason I used python 3 was because when I tried to install the unifi-portal with pip it threw an error; I forgot what it was specifically, but if I run into it today I'm sure I'll let you guys know.

**Comment by jMartich** [comment]
@KevSex 
My apologies for my busy days as well, I got the new VM going and I really do apologize in advance for my lack of understanding in setting up this project. So far I got python 2.7, django 1.11, then I cloned this project to ~/home/jmartich/Projects/Unifi-Portal.
So I have all the files for this project, but I can't seem to find the manage.py file. Based on the DJango documentation; django-admin startproject <projectname> will generate the manage.py file, then you can move on from there. 
My theory is that I have to run django-admin startproject django_unifi_portal, let my VM generate the manage.py file, merge this repo within my project, then run python manage.py. It's just a theory since I cannot seem to locate the manage.py file within this project. Please point out what I am doing wrong. You said you downloaded this repo, & I said I cloned; are those two different things or are they the same just different words? Once again please excuse my lack of experience with Github, I promise you I'll get better.

**Comment by KevSex** [comment]
@jMartich That's correct yeah! The startproject will create all the necessary files to create the default project. Once you've downloaded/cloned (same thing yes) this repo, the migrate command will essentially bring this into the project.

Some steps that I've been writing up since I started with a fresh install. Let us know if this helps...

Some prerequisites...

```
apt install python-pip virtualenv unzip
```

Setup our virtual environment...

```
mkdir /home/ubuntu/.virtualenvironments
cd /home/ubuntu/.virtualenvironments/
virtualenv unifi_portal
source unifi_portal/bin/activate
```

Install django...

```
pip install django==1.11.7
```

Install this portal...

```
pip install django-unifi-portal
```

Change to your chosen directory `~/home/jmartich/Projects/Unifi-Portal` and create the project...

```
django-admin startproject unifi
```

This will create all the necessary files therefore in your case, your manage.py file would be in `~/home/jmartich/Projects/Unifi-Portal/unifi/manage.py`
Change into the directory that was just created...

```
cd unifi
```

Verify you can run the project and visit the website...

```
python manage.py runserver 0:8000
```

Visit the site at http://public-ip:8000/. You should be presented with a `DisallowedHost at /` error with a yellow background (We'll correct this shortly).
_(Note: If the site isn't responding, check your security groups in AWS to ensure the port 8000 is open to your IP address.)_

Under unifi/settings.py, you'll need to update the line starting `ALLOWED_HOSTS` and add in the IP address of the VM in order to get rid of the yellow disallowed message.
After updating this file, re-running the project, and visiting the site again, you should then then see a blue background with a message ‘It Worked!’.

Next, download this repo...

```
wget https://github.com/bsab/django-unifi-portal/archive/master.zip
unzip master.zip
```

We don’t actually need all these files/directories, only the one.. `django_unifi_portal`. Move this and delete the rest.

```
mv django-unifi-portal-master/django_unifi_portal/ .
rm -rf django-unifi-portal-master/
rm -f master.zip
```

Next, we migrate this in to our project...

```
python manage.py migrate
```

Next, follow through steps 4 and 5 in the _Quick Start guide_ for creating unifi_settings.py and updating settings.py.
You'll need to update your UniFi details (server, port etc). 
_(Note: The UNIFI_SERVER is missing the end quotation " so make sure to add this in)._

Step 6 states to add in a line into url.py. Again, for you this would be in `~/home/jmartich/Projects/Unifi-Portal/unifi/unifi/url.py`. This line should be added within the ‘urlpatterns’ list.

```
url(r'', include('django_unifi_portal.urls')),
```

If we attempt to run the project just now however, it will complain about a NameError as include is not defined. So within this url.py file, you'll need to edit the first line to look like this...

```
from django.conf.urls import include, url
```

Finally, we can run the project again and attempt to visit the site...

```
python manage.py runserver 0:8000
```

The login page should then be accessible at `http://public-ip:8000/unifi-portal/login/`

Let us know how you get on :)

**Comment by jMartich** [comment]
@KevSex I appreciate all of your assistance, I'm going to give this a go later on today. What you wrote seems extremely detailed and made a lot of sense. Upon succeeding in this project, I will respond and close this issue; being that all my questions have been answered by you. Thanks again for all of your help and documentation.

**Comment by KevSex** [comment]
@jMartich How'd you get on with your django project? You get it up and running? Any errors in the steps?

**Comment by jMartich** [comment]
@KevSex, I will be doing this tomorrow. I did want to get it done on the day I mentioned prior. I will let you know if I run into any issues. Your instructions seem to have all the steps, I will verify with you tomorrow.

**Comment by jMartich** [comment]
@KevSex Alright, so my test setup was successful in a sense of deployment; I haven't added SMTP functionality or the FB API because this was just to get it running.
As for the instructions, I'm gonna give a you a A+. I got it up and running with no issues. The next part I need help with cannot be discussed here because it's off topic and Unifi Controller based. Here's some output for proof that I got it working:

[22/Dec/2017 18:43:44] "GET /static/material/fonts/material-design-icons/MaterialIcons-Regular.woff2 HTTP/1.1" 200 44300
[22/Dec/2017 18:45:30] "GET /unifi-portal/registration/ HTTP/1.1" 200 5731
Not Found: /unifi-portal/registration/<relative path under the static folder to the logo png>
WARNING:django.request:Not Found: /unifi-portal/registration/<relative path under the static folder to the logo png>
[22/Dec/2017 18:45:31] "GET /unifi-portal/registration/%3Crelative%20path%20under%20the%20static%20folder%20to%20the%20logo%20png%3E HTTP/1.1" 404 3118

I am aware of the errors, I consider this completed being that it was for a test server; its also up & running. Once again your instructions helped me a lot as a noob to django. 
Oh one more thing download & clone are different ;) wget vs git clone LOL. Thanks a bunch bud, it's been a great experience.
